### ISSUE TITLE
Failover taking too long

### ISSUE THREAD

**Issue by albionb96** [cloudnative-pg/cloudnative-pg#6154]
### Discussed in https://github.com/cloudnative-pg/cloudnative-pg/discussions/6146

<div type='discussions-op-text'>

<sup>Originally posted by **albionb96** November 21, 2024</sup>
Hi guys,

we were using another solution for our database until now, and lately we wanted to have a look at the coudnativePG solution.

The first thing that we tested was the failover time. 
We killed the Kubernetes node where the cluster primary was living in order to simulate a node failure, and we observed from a client for how long this client cannot send inserts into db.

With our old solution we had a failover time between 25-35 seconds (during this time the application was not able to insert data  into db).

When executing the same failover tests with cloudnativePG we got a failover time somewhere between 40-80seconds.
(If the controller manager pod would be living in the same node like the primary this time could be much longer.)

What we noticed is that the WAL receivers are timed out exactly after 5seconds but it takes some 30-37 additional seconds  (which sums up to 35-42 seconds after the node failure, if we calculate the 5s timeout we had for wal receivers) for the cluster to even just start the failover process in the first place. At least we do not see anything in the logs related to the failover during this time. And although we tried to modify some of the config parameters we didn't achieve better results.

I thought to myself that maybe this time might be impacted from something hardcoded, and after looking a little bit at the source code I doubt that this time is coming from the cluster status timeout defined here, but I am not a golang dev or expert, so I don't know:

![image](https://github.com/user-attachments/assets/3bc3fcbc-04b4-45d8-a4ba-462674472fc2)
https://github.com/cloudnative-pg/cloudnative-pg/blob/72e2f5c490fd17346f1f9ff539589a8d9f01690c/internal/cmd/manager/instance/status/cmd.go#L121

Somewhere I read that in case of node failure the cloudnativepg looks at the node status if it is not-ready or unreachable, and then based on tolerations decides to kill the primary pod and start the failover process. If that is true I now that by default the node-monitor-grace-period is 40s that means that it takes 40s for the kube controller manager to mark node as unhealthy, and that config cannot be adapted. If that is true why is the cloudnativePG relying on that and not triggering the failover process based on its cluster status results/timeouts?

My questions is: Is it possible to reduce somehow the failover time when the node of primary is killed or fails? Can we achieve the old range we had 25-35 seconds? Any idea?

Some more details:

Below we see the logs from the client which was not able to write to db for around 51 seconds during the failover:

`Connection is Valid: 2024-11-19T15:02:17.141189046+01:00Sql: INSERT INTO time (time) VALUES (to_timestamp('2024-11-19T15:02:17.141189046+01:00', 'YYYY-MM-DDThh24:mi:ss')::timestamp with time zone at time zone 'Etc/UTC');
Connection is Valid: **2024-11-19T15:02:18.141194779+01:00**Sql: INSERT INTO time (time) VALUES (to_timestamp('2024-11-19T15:02:18.141194779+01:00', 'YYYY-MM-DDThh24:mi:ss')::timestamp with time zone at time zone 'Etc/UTC');
2024-11-19T14:02:29.150Z  WARN 1 --- [   scheduling-1] com.zaxxer.hikari.pool.PoolBase          : HikariPool-1 - Failed to validate connection org.postgresql.jdbc.PgConnection@3fc0c66f (This connection has been closed.). Possibly consider using a shorter maxLifetime value.
Problem: HikariPool-1 - Connection is not available, request timed out after 10009ms (total=0, active=0, idle=0, waiting=0), Failure: 2024-11-19T15:02:19.141219498+01:00
Problem: HikariPool-1 - Connection is not available, request timed out after 2000ms (total=0, active=0, idle=0, waiting=0), Failure: 2024-11-19T15:02:29.155141629+01:00
Problem: HikariPool-1 - Connection is not available, request timed out after 2000ms (total=0, active=0, idle=0, waiting=0), Failure: 2024-11-19T15:02:31.156196969+01:00
Problem: HikariPool-1 - Connection is not available, request timed out after 2000ms (total=0, active=0, idle=0, waiting=0), Failure: 2024-11-19T15:02:33.156514975+01:00
Problem: HikariPool-1 - Connection is not available, request timed out after 2000ms (total=0, active=0, idle=0, waiting=0), Failure: 2024-11-19T15:02:35.156847851+01:00
Problem: HikariPool-1 - Connection is not available, request timed out after 2000ms (total=0, active=0, idle=0, waiting=0), Failure: 2024-11-19T15:02:37.157299363+01:00
Problem: HikariPool-1 - Connection is not available, request timed out after 2000ms (total=0, active=0, idle=0, waiting=0), Failure: 2024-11-19T15:02:39.157611065+01:00
Problem: HikariPool-1 - Connection is not available, request timed out after 2000ms (total=0, active=0, idle=0, waiting=0), Failure: 2024-11-19T15:02:41.157907340+01:00
Problem: HikariPool-1 - Connection is not available, request timed out after 2000ms (total=0, active=0, idle=0, waiting=0), Failure: 2024-11-19T15:02:43.158215925+01:00
Problem: HikariPool-1 - Connection is not available, request timed out after 2000ms (total=0, active=0, idle=0, waiting=0), Failure: 2024-11-19T15:02:45.158519959+01:00
Problem: HikariPool-1 - Connection is not available, request timed out after 2000ms (total=0, active=0, idle=0, waiting=0), Failure: 2024-11-19T15:02:47.158822915+01:00
Problem: HikariPool-1 - Connection is not available, request timed out after 2000ms (total=0, active=0, idle=0, waiting=0), Failure: 2024-11-19T15:02:49.159112575+01:00
Problem: HikariPool-1 - Connection is not available, request timed out after 2000ms (total=0, active=0, idle=0, waiting=0), Failure: 2024-11-19T15:02:51.159612646+01:00
Problem: HikariPool-1 - Connection is not available, request timed out after 2000ms (total=0, active=0, idle=0, waiting=0), Failure: 2024-11-19T15:02:53.159916412+01:00
Problem: HikariPool-1 - Connection is not available, request timed out after 2000ms (total=0, active=0, idle=0, waiting=0), Failure: 2024-11-19T15:02:55.160769376+01:00
Problem: HikariPool-1 - Connection is not available, request timed out after 2000ms (total=0, active=0, idle=0, waiting=0), Failure: 2024-11-19T15:02:57.161084043+01:00
Problem: HikariPool-1 - Connection is not available, request timed out after 2000ms (total=0, active=0, idle=0, waiting=0), Failure: 2024-11-19T15:02:59.161413739+01:00
Problem: HikariPool-1 - Connection is not available, request timed out after 2000ms (total=0, active=0, idle=0, waiting=0), Failure: 2024-11-19T15:03:01.161729031+01:00
Problem: HikariPool-1 - Connection is not available, request timed out after 2000ms (total=0, active=0, idle=0, waiting=0), Failure: 2024-11-19T15:03:03.162039029+01:00
Problem: HikariPool-1 - Connection is not available, request timed out after 2000ms (total=0, active=0, idle=0, waiting=0), Failure: 2024-11-19T15:03:05.162364989+01:00
Problem: HikariPool-1 - Connection is not available, request timed out after 2000ms (total=0, active=0, idle=0, waiting=0), Failure: 2024-11-19T15:03:07.162657327+01:00
Connection is Valid: **2024-11-19T15:03:09.162953329+01:00**Sql: INSERT INTO time (time) VALUES (to_timestamp('2024-11-19T15:03:09.162953329+01:00', 'YYYY-MM-DDThh24:mi:ss')::timestamp with time zone at time zone 'Etc/UTC');
Outage duration: 51
Connection is Valid: 2024-11-19T15:03:09.331016425+01:00Sql: INSERT INTO time (time) VALUES (to_timestamp('2024-11-19T15:03:09.331016425+01:00', 'YYYY-MM-DDThh24:mi:ss')::timestamp with time zone at time zone 'Etc/UTC');
`

Below we see the logs of the cluster replica which will be elected as primary after the failover.
We see that the wal receivers started to terminate 5seconds after the failure but it took around 37 additional seconds too see some further logs regarding the failover process itself:

`{"level":"info","ts":"`**2024-11-19T14:02:23.147275668**Z`","logger":"postgres","msg":"record","logging_pod":"cluster-sample-1","record":{"log_time":"2024-11-19 14:02:23.147 UTC","process_id":"130","session_id":"673c9568.82","session_line_num":"2","session_start_time":"2024-11-19 13:40:56 UTC","transaction_id":"0","error_severity":"FATAL","sql_state_code":"08006","message":**"terminating walreceiver** due to timeout","backend_type":"walreceiver","query_id":"0"}}
{"level":"debug","ts":"2024-11-19T14:02:23.172584863Z","msg":"Cached object request received","caller":"pkg/mod/github.com/cloudnative-pg/machinery@v0.0.0-20241014090714-c27747f9974b/pkg/log/log.go:178","logging_pod":"cluster-sample-1"}
{"level":"debug","ts":"2024-11-19T14:02:23.303569229Z","msg":"Cached object request received","caller":"pkg/mod/github.com/cloudnative-pg/machinery@v0.0.0-20241014090714-c27747f9974b/pkg/log/log.go:178","logging_pod":"cluster-sample-1"}
{"level":"info","ts":"2024-11-19T14:02:23.408597375Z","logger":"postgres","msg":"record","logging_pod":"cluster-sample-1","record":{"log_time":"2024-11-19 14:02:23.408 UTC","process_id":"43","session_id":"673c9567.2b","session_line_num":"8","session_start_time":"2024-11-19 13:40:55 UTC","virtual_transaction_id":"148/0","transaction_id":"0","error_severity":"LOG","sql_state_code":"00000","message":"invalid record length at 0/E0083D8: expected at least 24, got 0","backend_type":"startup","query_id":"0"}}
{"level":"info","ts":"2024-11-19T14:02:58.375493965Z","logger":"Replicator","msg":"synchronizing replication slots","logger":"instance-manager","logging_pod":"cluster-sample-1","err":"getting replication slot status from primary: failed to connect to `user=streaming_replica database=postgres`: 172.30.33.8:5432 (cluster-sample-rw): dial error: dial tcp 172.30.33.8:5432: connect: connection refused"}
{"level":"info","ts":"2024-11-19T14:02:58.376720004Z","logger":"Replicator","msg":"synchronizing replication slots","logger":"instance-manager","logging_pod":"cluster-sample-1","err":"getting replication slot status from primary: failed to connect to `user=streaming_replica database=postgres`: 172.30.33.8:5432 (cluster-sample-rw): dial error: dial tcp 172.30.33.8:5432: connect: connection refused"}
{"level":"debug","ts":"**2024-11-19T14:03:00.430552602Z**","logger":"external_servers_reconciler","msg":"starting up the external servers reconciler","logger":"instance-manager","logging_pod":"cluster-sample-1","controller":"instance-external-server","controllerGroup":"postgresql.cnpg.io","controllerKind":"Cluster","Cluster":{"name":"cluster-sample","namespace":"openshift-operators"},"namespace":"openshift-operators","name":"cluster-sample","reconcileID":"575b16c0-d8c4-48a1-b3af-c29fff36cf1c","caller":"internal/management/controller/externalservers/reconciler.go:56","logging_pod":"cluster-sample-1"}
{"level":"debug","ts":"2024-11-19T14:03:00.430567114Z","msg":"Reconciling Cluster","logger":"instance-manager","logging_pod":"cluster-sample-1","controller":"instance-cluster","controllerGroup":"postgresql.cnpg.io","controllerKind":"Cluster","Cluster":{"name":"cluster-sample","namespace":"openshift-operators"},"namespace":"openshift-operators","name":"cluster-sample","reconcileID":"5b2e572b-ae50-4134-9c2f-0844f9f2ceda","caller":"internal/management/controller/instance_controller.go:115","logging_pod":"cluster-sample-1","cluster":{"apiVersion":"postgresql.cnpg.io/v1","kind":"Cluster","namespace":"openshift-operators","name":"cluster-sample"}}
{"level":"debug","ts":"2024-11-19T14:03:00.430617394Z","msg":"Reconciling custom monitoring queries","logger":"instance-manager","logging_pod":"cluster-sample-1","controller":"instance-cluster","controllerGroup":"postgresql.cnpg.io","controllerKind":"Cluster","Cluster":{"name":"cluster-sample","namespace":"openshift-operators"},"namespace":"openshift-operators","name":"cluster-sample","reconcileID":"5b2e572b-ae50-4134-9c2f-0844f9f2ceda","caller":"internal/management/controller/instance_controller.go:807","logging_pod":"cluster-sample-1"}
{"level":"debug","ts":"2024-11-19T14:03:00.430674265Z","logger":"tbs_reconciler","msg":"skipping the tablespace reconciler in replicas","logger":"instance-manager","logging_pod":"cluster-sample-1","controller":"instance-tablespaces","controllerGroup":"postgresql.cnpg.io","controllerKind":"Cluster","Cluster":{"name":"cluster-sample","namespace":"openshift-operators"},"namespace":"openshift-operators","name":"cluster-sample","reconcileID":"bda1796f-7c85-407d-bf7d-3b9f7a6cf9b5","caller":"internal/management/controller/tablespaces/reconciler.go:52","logging_pod":"cluster-sample-1"}
{"level":"debug","ts":"2024-11-19T14:03:00.440898957Z","logger":"tbs_reconciler","msg":"skipping the tablespace reconciler in replicas","logger":"instance-manager","logging_pod":"cluster-sample-1","controller":"instance-tablespaces","controllerGroup":"postgresql.cnpg.io","controllerKind":"Cluster","Cluster":{"name":"cluster-sample","namespace":"openshift-operators"},"namespace":"openshift-operators","name":"cluster-sample","reconcileID":"4f42d5ab-b60c-4025-8837-f959b277137d","caller":"internal/management/controller/tablespaces/reconciler.go:52","logging_pod":"cluster-sample-1"}
{"level":"debug","ts":"2024-11-19T14:03:00.440916345Z","logger":"external_servers_reconciler","msg":"starting up the external servers reconciler","logger":"instance-manager","logging_pod":"cluster-sample-1","controller":"instance-external-server","controllerGroup":"postgresql.cnpg.io","controllerKind":"Cluster","Cluster":{"name":"cluster-sample","namespace":"openshift-operators"},"namespace":"openshift-operators","name":"cluster-sample","reconcileID":"94d16774-2ee8-4d11-97a6-1b6c42da40a5","caller":"internal/management/controller/externalservers/reconciler.go:56","logging_pod":"cluster-sample-1"}
{"level":"debug","ts":"2024-11-19T14:03:00.455005845Z","logger":"external_servers_reconciler","msg":"starting up the external servers reconciler","logger":"instance-manager","logging_pod":"cluster-sample-1","controller":"instance-external-server","controllerGroup":"postgresql.cnpg.io","controllerKind":"Cluster","Cluster":{"name":"cluster-sample","namespace":"openshift-operators"},"namespace":"openshift-operators","name":"cluster-sample","reconcileID":"9f1e441b-8e61-4364-94a4-2c0b5d24bc5c","caller":"internal/management/controller/externalservers/reconciler.go:56","logging_pod":"cluster-sample-1"}
{"level":"debug","ts":"2024-11-19T14:03:00.455086917Z","logger":"tbs_reconciler","msg":"skipping the tablespace reconciler in replicas","logger":"instance-manager","logging_pod":"cluster-sample-1","controller":"instance-tablespaces","controllerGroup":"postgresql.cnpg.io","controllerKind":"Cluster","Cluster":{"name":"cluster-sample","namespace":"openshift-operators"},"namespace":"openshift-operators","name":"cluster-sample","reconcileID":"0000e57f-6943-453b-8239-bb9dd88ba92b","caller":"internal/management/controller/tablespaces/reconciler.go:52","logging_pod":"cluster-sample-1"}
{"level":"debug","ts":"2024-11-19T14:03:00.464739371Z","logger":"tbs_reconciler","msg":"skipping the tablespace reconciler in replicas","logger":"instance-manager","logging_pod":"cluster-sample-1","controller":"instance-tablespaces","controllerGroup":"postgresql.cnpg.io","controllerKind":"Cluster","Cluster":{"name":"cluster-sample","namespace":"openshift-operators"},"namespace":"openshift-operators","name":"cluster-sample","reconcileID":"e1bdbf8f-4c97-41eb-b7f1-98348b7eaa8b","caller":"internal/management/controller/tablespaces/reconciler.go:52","logging_pod":"cluster-sample-1"}
{"level":"debug","ts":"2024-11-19T14:03:00.46486145Z","logger":"external_servers_reconciler","msg":"starting up the external servers reconciler","logger":"instance-manager","logging_pod":"cluster-sample-1","controller":"instance-external-server","controllerGroup":"postgresql.cnpg.io","controllerKind":"Cluster","Cluster":{"name":"cluster-sample","namespace":"openshift-operators"},"namespace":"openshift-operators","name":"cluster-sample","reconcileID":"e830b45c-fcdf-4214-890a-da3429522b03","caller":"internal/management/controller/externalservers/reconciler.go:56","logging_pod":"cluster-sample-1"}
{"level":"debug","ts":"2024-11-19T14:03:00.476379816Z","logger":"tbs_reconciler","msg":"skipping the tablespace reconciler in replicas","logger":"instance-manager","logging_pod":"cluster-sample-1","controller":"instance-tablespaces","controllerGroup":"postgresql.cnpg.io","controllerKind":"Cluster","Cluster":{"name":"cluster-sample","namespace":"openshift-operators"},"namespace":"openshift-operators","name":"cluster-sample","reconcileID":"4ca5d71d-f9b4-4a3a-aa65-9e75d819ba6f","caller":"internal/management/controller/tablespaces/reconciler.go:52","logging_pod":"cluster-sample-1"}
{"level":"debug","ts":"2024-11-19T14:03:00.476433933Z","logger":"external_servers_reconciler","msg":"starting up the external servers reconciler","logger":"instance-manager","logging_pod":"cluster-sample-1","controller":"instance-external-server","controllerGroup":"postgresql.cnpg.io","controllerKind":"Cluster","Cluster":{"name":"cluster-sample","namespace":"openshift-operators"},"namespace":"openshift-operators","name":"cluster-sample","reconcileID":"3e0bf7f3-9a00-4043-8b79-590f86b500d2","caller":"internal/management/controller/externalservers/reconciler.go:56","logging_pod":"cluster-sample-1"}
{"level":"debug","ts":"2024-11-19T14:03:00.495962846Z","msg":"Reconciling Cluster","logger":"instance-manager","logging_pod":"cluster-sample-1","controller":"instance-cluster","controllerGroup":"postgresql.cnpg.io","controllerKind":"Cluster","Cluster":{"name":"cluster-sample","namespace":"openshift-operators"},"namespace":"openshift-operators","name":"cluster-sample","reconcileID":"4a2d265e-5a49-458b-bae6-f1dd8cc957ff","caller":"internal/management/controller/instance_controller.go:115","logging_pod":"cluster-sample-1","cluster":{"apiVersion":"postgresql.cnpg.io/v1","kind":"Cluster","namespace":"openshift-operators","name":"cluster-sample"}}
{"level":"debug","ts":"2024-11-19T14:03:00.496001228Z","msg":"Reconciling custom monitoring queries","logger":"instance-manager","logging_pod":"cluster-sample-1","controller":"instance-cluster","controllerGroup":"postgresql.cnpg.io","controllerKind":"Cluster","Cluster":{"name":"cluster-sample","namespace":"openshift-operators"},"namespace":"openshift-operators","name":"cluster-sample","reconcileID":"4a2d265e-5a49-458b-bae6-f1dd8cc957ff","caller":"internal/management/controller/instance_controller.go:807","logging_pod":"cluster-sample-1"}
{"level":"info","ts":"2024-11-19T14:03:00.497249192Z","logger":"Replicator","msg":"synchronizing replication slots","logger":"instance-manager","logging_pod":"cluster-sample-1","err":"getting replication slot status from primary: failed to connect to `user=streaming_replica database=postgres`: 172.30.33.8:5432 (cluster-sample-rw): dial error: dial tcp 172.30.33.8:5432: connect: connection refused"}
{"level":"info","ts":"2024-11-19T14:03:00.551899561Z","msg":"Setting myself as primary","logger":"instance-manager","logging_pod":"cluster-sample-1","controller":"instance-cluster","controllerGroup":"postgresql.cnpg.io","controllerKind":"Cluster","Cluster":{"name":"cluster-sample","namespace":"openshift-operators"},"namespace":"openshift-operators","name":"cluster-sample","reconcileID":"4a2d265e-5a49-458b-bae6-f1dd8cc957ff","logging_pod":"cluster-sample-1","phase":"Failing over","currentTimestamp":"2024-11-19T14:03:00.551882Z","targetPrimaryTimestamp":"2024-11-19T14:03:00.463507Z","currentPrimaryTimestamp":"2024-11-19T13:35:09.903398Z","msPassedSinceTargetPrimaryTimestamp":88,"msPassedSinceCurrentPrimaryTimestamp":1670648,"msDifferenceBetweenCurrentAndTargetPrimary":-1670560}
{"level":"info","ts":"2024-11-19T14:03:00.551924762Z","msg":"I'm the target primary, wait for the wal_receiver to be terminated","logger":"instance-manager","logging_pod":"cluster-sample-1","controller":"instance-cluster","controllerGroup":"postgresql.cnpg.io","controllerKind":"Cluster","Cluster":{"name":"cluster-sample","namespace":"openshift-operators"},"namespace":"openshift-operators","name":"cluster-sample","reconcileID":"4a2d265e-5a49-458b-bae6-f1dd8cc957ff","logging_pod":"cluster-sample-1"}
{"level":"info","ts":"2024-11-19T14:03:00.554814361Z","msg":"I'm the target primary, applying WALs and promoting my instance","logger":"instance-manager","logging_pod":"cluster-sample-1","controller":"instance-cluster","controllerGroup":"postgresql.cnpg.io","controllerKind":"Cluster","Cluster":{"name":"cluster-sample","namespace":"openshift-operators"},"namespace":"openshift-operators","name":"cluster-sample","reconcileID":"4a2d265e-5a49-458b-bae6-f1dd8cc957ff","logging_pod":"cluster-sample-1"}
{"level":"info","ts":"2024-11-19T14:03:00.554833716Z","msg":"Extracting pg_controldata information","logger":"instance-manager","logging_pod":"cluster-sample-1","controller":"instance-cluster","controllerGroup":"postgresql.cnpg.io","controllerKind":"Cluster","Cluster":{"name":"cluster-sample","namespace":"openshift-operators"},"namespace":"openshift-operators","name":"cluster-sample","reconcileID":"4a2d265e-5a49-458b-bae6-f1dd8cc957ff","logging_pod":"cluster-sample-1","reason":"promote"}
{"level":"info","ts":"2024-11-19T14:03:00.555859512Z","logger":"pg_controldata","msg":"pg_control version number:            1700\nCatalog version number:               202406281\nDatabase system identifier:           7438970221660221468\nDatabase cluster state:               in archive recovery\npg_control last modified:             Tue 19 Nov 2024 02:00:56 PM UTC\nLatest checkpoint location:           0/E002458\nLatest checkpoint's REDO location:    0/D0150D8\nLatest checkpoint's REDO WAL file:    00000002000000000000000D\nLatest checkpoint's TimeLineID:       2\nLatest checkpoint's PrevTimeLineID:   2\nLatest checkpoint's full_page_writes: on\nLatest checkpoint's NextXID:          0:2380\nLatest checkpoint's NextOID:          24578\nLatest checkpoint's NextMultiXactId:  1\nLatest checkpoint's NextMultiOffset:  0\nLatest checkpoint's oldestXID:        730\nLatest checkpoint's oldestXID's DB:   1\nLatest checkpoint's oldestActiveXID:  2380\nLatest checkpoint's oldestMultiXid:   1\nLatest checkpoint's oldestMulti's DB: 1\nLatest checkpoint's oldestCommitTsXid:0\nLatest checkpoint's newestCommitTsXid:0\nTime of latest checkpoint:            Tue 19 Nov 2024 02:00:09 PM UTC\nFake LSN counter for unlogged rels:   0/3E8\nMinimum recovery ending location:     0/E004618\nMin recovery ending loc's timeline:   2\nBackup start location:                0/0\nBackup end location:                  0/0\nEnd-of-backup record required:        no\nwal_level setting:                    logical\nwal_log_hints setting:                on\nmax_connections setting:              100\nmax_worker_processes setting:         32\nmax_wal_senders setting:              10\nmax_prepared_xacts setting:           0\nmax_locks_per_xact setting:           64\ntrack_commit_timestamp setting:       off\nMaximum data alignment:               8\nDatabase block size:                  8192\nBlocks per segment of large relation: 131072\nWAL block size:                       8192\nBytes per WAL segment:                16777216\nMaximum length of identifiers:        64\nMaximum columns in an index:          32\nMaximum size of a TOAST chunk:        1996\nSize of a large-object chunk:         2048\nDate/time type storage:               64-bit integers\nFloat8 argument passing:              by value\nData page checksum version:           0\nMock authentication nonce:            394ee91a692b830e68552a0f1ced84562d6115a83a3d2179e7ca9d5f34dffb03\n","pipe":"stdout","logging_pod":"cluster-sample-1"}
{"level":"info","ts":"2024-11-19T14:03:00.555881163Z","msg":"Promoting instance","logger":"instance-manager","logging_pod":"cluster-sample-1","controller":"instance-cluster","controllerGroup":"postgresql.cnpg.io","controllerKind":"Cluster","Cluster":{"name":"cluster-sample","namespace":"openshift-operators"},"namespace":"openshift-operators","name":"cluster-sample","reconcileID":"4a2d265e-5a49-458b-bae6-f1dd8cc957ff","logging_pod":"cluster-sample-1","pgctl_options":["-D","/var/lib/postgresql/data/pgdata","-w","promote","-t 40000000"]}
`

Below we see the logs from the controller manager, which pod doesn't log anything at all during the first 42 seconds, after the node failure. It starts logging something about the failover process at the same time as the logs in the replica that we have seen.

`
{"level":"info","ts":"2024-11-19T13:40:57.085473677Z","msg":"Cluster is healthy","controller":"cluster","controllerGroup":"postgresql.cnpg.io","controllerKind":"Cluster","Cluster":{"name":"cluster-sample","namespace":"openshift-operators"},"namespace":"openshift-operators","name":"cluster-sample","reconcileID":"cc46950c-dbd8-4ff7-9473-87c616ad1206"}
{"level":"info","ts":"**2024-11-19T14:03:00.419148934Z**","msg":"Cannot extract Pod status","controller":"cluster","controllerGroup":"postgresql.cnpg.io","controllerKind":"Cluster","Cluster":{"name":"cluster-sample","namespace":"openshift-operators"},"namespace":"openshift-operators","name":"cluster-sample","reconcileID":"a773349e-854e-4e7a-959d-c2c667e4ca1d","name":"cluster-sample-2","error":"Get \"https://10.244.5.186:8000/pg/status\": dial tcp 10.244.5.186:8000: i/o timeout"}
{"level":"info","ts":"2024-11-19T14:03:00.429902684Z","msg":"Current primary isn't healthy, initiating a failover","controller":"cluster","controllerGroup":"postgresql.cnpg.io","controllerKind":"Cluster","Cluster":{"name":"cluster-sample","namespace":"openshift-operators"},"namespace":"openshift-operators","name":"cluster-sample","reconcileID":"a773349e-854e-4e7a-959d-c2c667e4ca1d"}
{"level":"info","ts":"2024-11-19T14:03:00.42992148Z","msg":"pod status (1 of 3)","controller":"cluster","controllerGroup":"postgresql.cnpg.io","controllerKind":"Cluster","Cluster":{"name":"cluster-sample","namespace":"openshift-operators"},"namespace":"openshift-operators","name":"cluster-sample","reconcileID":"a773349e-854e-4e7a-959d-c2c667e4ca1d","name":"cluster-sample-1","currentLsn":"","receivedLsn":"0/E0083D8","replayLsn":"0/E0083D8","isPrimary":false,"isPodReady":true,"pendingRestart":false,"pendingRestartForDecrease":false,"statusCollectionError":null}
{"level":"info","ts":"2024-11-19T14:03:00.429937552Z","msg":"pod status (2 of 3)","controller":"cluster","controllerGroup":"postgresql.cnpg.io","controllerKind":"Cluster","Cluster":{"name":"cluster-sample","namespace":"openshift-operators"},"namespace":"openshift-operators","name":"cluster-sample","reconcileID":"a773349e-854e-4e7a-959d-c2c667e4ca1d","name":"cluster-sample-3","currentLsn":"","receivedLsn":"0/E0083D8","replayLsn":"0/E0083D8","isPrimary":false,"isPodReady":true,"pendingRestart":false,"pendingRestartForDecrease":false,"statusCollectionError":null}
{"level":"info","ts":"2024-11-19T14:03:00.429944256Z","msg":"pod status (3 of 3)","controller":"cluster","controllerGroup":"postgresql.cnpg.io","controllerKind":"Cluster","Cluster":{"name":"cluster-sample","namespace":"openshift-operators"},"namespace":"openshift-operators","name":"cluster-sample","reconcileID":"a773349e-854e-4e7a-959d-c2c667e4ca1d","name":"cluster-sample-2","currentLsn":"","receivedLsn":"","replayLsn":"","isPrimary":false,"isPodReady":true,"pendingRestart":false,"pendingRestartForDecrease":false,"statusCollectionError":"Get \"https://10.244.5.186:8000/pg/status\": dial tcp 10.244.5.186:8000: i/o timeout"}
{"level":"info","ts":"2024-11-19T14:03:00.430038243Z","msg":"Cluster is not healthy","controller":"cluster","controllerGroup":"postgresql.cnpg.io","controllerKind":"Cluster","Cluster":{"name":"cluster-sample","namespace":"openshift-operators"},"namespace":"openshift-operators","name":"cluster-sample","reconcileID":"a773349e-854e-4e7a-959d-c2c667e4ca1d"}
{"level":"info","ts":"2024-11-19T14:03:00.45246805Z","msg":"Failing over","controller":"cluster","controllerGroup":"postgresql.cnpg.io","controllerKind":"Cluster","Cluster":{"name":"cluster-sample","namespace":"openshift-operators"},"namespace":"openshift-operators","name":"cluster-sample","reconcileID":"a773349e-854e-4e7a-959d-c2c667e4ca1d","newPrimary":"cluster-sample-1"}
{"level":"info","ts":"2024-11-19T14:03:00.452497233Z","msg":"pod status (1 of 3)","controller":"cluster","controllerGroup":"postgresql.cnpg.io","controllerKind":"Cluster","Cluster":{"name":"cluster-sample","namespace":"openshift-operators"},"namespace":"openshift-operators","name":"cluster-sample","reconcileID":"a773349e-854e-4e7a-959d-c2c667e4ca1d","name":"cluster-sample-1","currentLsn":"","receivedLsn":"0/E0083D8","replayLsn":"0/E0083D8","isPrimary":false,"isPodReady":true,"pendingRestart":false,"pendingRestartForDecrease":false,"statusCollectionError":null}
{"level":"info","ts":"2024-11-19T14:03:00.452513933Z","msg":"pod status (2 of 3)","controller":"cluster","controllerGroup":"postgresql.cnpg.io","controllerKind":"Cluster","Cluster":{"name":"cluster-sample","namespace":"openshift-operators"},"namespace":"openshift-operators","name":"cluster-sample","reconcileID":"a773349e-854e-4e7a-959d-c2c667e4ca1d","name":"cluster-sample-3","currentLsn":"","receivedLsn":"0/E0083D8","replayLsn":"0/E0083D8","isPrimary":false,"isPodReady":true,"pendingRestart":false,"pendingRestartForDecrease":false,"statusCollectionError":null}
{"level":"info","ts":"2024-11-19T14:03:00.452520661Z","msg":"pod status (3 of 3)","controller":"cluster","controllerGroup":"postgresql.cnpg.io","controllerKind":"Cluster","Cluster":{"name":"cluster-sample","namespace":"openshift-operators"},"namespace":"openshift-operators","name":"cluster-sample","reconcileID":"a773349e-854e-4e7a-959d-c2c667e4ca1d","name":"cluster-sample-2","currentLsn":"","receivedLsn":"","replayLsn":"","isPrimary":false,"isPodReady":true,"pendingRestart":false,"pendingRestartForDecrease":false,"statusCollectionError":"Get \"https://10.244.5.186:8000/pg/status\": dial tcp 10.244.5.186:8000: i/o timeout"}
{"level":"info","ts":"2024-11-19T14:03:00.475254228Z","msg":"Waiting for the new primary to notice the promotion request","controller":"cluster","controllerGroup":"postgresql.cnpg.io","controllerKind":"Cluster","Cluster":{"name":"cluster-sample","namespace":"openshift-operators"},"namespace":"openshift-operators","name":"cluster-sample","reconcileID":"a773349e-854e-4e7a-959d-c2c667e4ca1d","newPrimary":"cluster-sample-1"}
{"level":"info","ts":"2024-11-19T14:03:00.537044226Z","msg":"There is a switchover or a failover in progress, waiting for the operation to complete","controller":"cluster","controllerGroup":"postgresql.cnpg.io","controllerKind":"Cluster","Cluster":{"name":"cluster-sample","namespace":"openshift-operators"},"namespace":"openshift-operators","name":"cluster-sample","reconcileID":"470d2f40-97a4-4262-90ea-d75074d539d2","currentPrimary":"cluster-sample-2","targetPrimary":"cluster-sample-1"}
{"level":"info","ts":"2024-11-19T14:03:00.592989434Z","msg":"There is a switchover or a failover in progress, waiting for the operation to complete","controller":"cluster","controllerGroup":"postgresql.cnpg.io","controllerKind":"Cluster","Cluster":{"name":"cluster-sample","namespace":"openshift-operators"},"namespace":"openshift-operators","name":"cluster-sample","reconcileID":"78b742d6-3da3-4ed1-9fb2-1d270e23bdb2","currentPrimary":"cluster-sample-2","targetPrimary":"cluster-sample-1"}
{"level":"info","ts":"2024-11-19T14:03:03.575714289Z","msg":"Cannot extract Pod status","controller":"cluster","controllerGroup":"postgresql.cnpg.io","controllerKind":"Cluster","Cluster":{"name":"cluster-sample","namespace":"openshift-operators"},"namespace":"openshift-operators","name":"cluster-sample","reconcileID":"a92cad56-4eec-4a17-9f55-9b5f8ebf59cc","name":"cluster-sample-2","error":"Get \"https://10.244.5.186:8000/pg/status\": dial tcp 10.244.5.186:8000: i/o timeout"}
{"level":"info","ts":"2024-11-19T14:03:05.708673906Z","msg":"Cannot extract Pod status","controller":"cluster","controllerGroup":"postgresql.cnpg.io","controllerKind":"Cluster","Cluster":{"name":"cluster-sample","namespace":"openshift-operators"},"namespace":"openshift-operators","name":"cluster-sample","reconcileID":"065e7936-51e9-4979-a4a9-4eff0cc315bc","name":"cluster-sample-2","error":"Get \"https://10.244.5.186:8000/pg/status\": dial tcp 10.244.5.186:8000: i/o timeout"}
{"level":"info","ts":"2024-11-19T14:03:05.718818785Z","msg":"Updating pvc metadata","controller":"cluster","controllerGroup":"postgresql.cnpg.io","controllerKind":"Cluster","Cluster":{"name":"cluster-sample","namespace":"openshift-operators"},"namespace":"openshift-operators","name":"cluster-sample","reconcileID":"065e7936-51e9-4979-a4a9-4eff0cc315bc","pvc":"cluster-sample-1","reconciler":"instance-role"}
{"level":"info","ts":"2024-11-19T14:03:05.727137941Z","msg":"Updating pvc metadata","controller":"cluster","controllerGroup":"postgresql.cnpg.io","controllerKind":"Cluster","Cluster":{"name":"cluster-sample","namespace":"openshift-operators"},"namespace":"openshift-operators","name":"cluster-sample","reconcileID":"065e7936-51e9-4979-a4a9-4eff0cc315bc","pvc":"cluster-sample-1-wal","reconciler":"instance-role"}
{"level":"info","ts":"2024-11-19T14:03:05.735681781Z","msg":"Updating pvc metadata","controller":"cluster","controllerGroup":"postgresql.cnpg.io","controllerKind":"Cluster","Cluster":{"name":"cluster-sample","namespace":"openshift-operators"},"namespace":"openshift-operators","name":"cluster-sample","reconcileID":"065e7936-51e9-4979-a4a9-4eff0cc315bc","pvc":"cluster-sample-2","reconciler":"instance-role"}
{"level":"info","ts":"2024-11-19T14:03:05.747854475Z","msg":"Updating pvc metadata","controller":"cluster","controllerGroup":"postgresql.cnpg.io","controllerKind":"Cluster","Cluster":{"name":"cluster-sample","namespace":"openshift-operators"},"namespace":"openshift-operators","name":"cluster-sample","reconcileID":"065e7936-51e9-4979-a4a9-4eff0cc315bc","pvc":"cluster-sample-2-wal","reconciler":"instance-role"}
{"level":"info","ts":"2024-11-19T14:03:05.753609429Z","msg":"Setting primary label","controller":"cluster","controllerGroup":"postgresql.cnpg.io","controllerKind":"Cluster","Cluster":{"name":"cluster-sample","namespace":"openshift-operators"},"namespace":"openshift-operators","name":"cluster-sample","reconcileID":"065e7936-51e9-4979-a4a9-4eff0cc315bc","pod":"cluster-sample-1"}
{"level":"info","ts":"2024-11-19T14:03:05.76791717Z","msg":"Setting replica label","controller":"cluster","controllerGroup":"postgresql.cnpg.io","controllerKind":"Cluster","Cluster":{"name":"cluster-sample","namespace":"openshift-operators"},"namespace":"openshift-operators","name":"cluster-sample","reconcileID":"065e7936-51e9-4979-a4a9-4eff0cc315bc","pod":"cluster-sample-2"}
`</div>

**Comment by gbartolini** [comment]
Thank you for opening this ticket and for your time spent investigating. We will look into this and check all timeouts.

FWIW, we highly discourage running the controller on the same nodes where Postgres runs. Postgres should be isolated from the rest, and proper anti-affinity rules should be in place: https://cloudnative-pg.io/documentation/current/architecture/#reserving-nodes-for-postgresql-workloads

**Comment by albionb96** [comment]
@gbartolini Thank you for your reply. If you need anything else from me just let me know. 

Actually putting the controller in the same node as a postgres workloads was only an early experiment, all the later failovertests were executed, when the controller pod was deployed in another node.

Bur are you saying that not only the controller pod, but also all workloads other than the postgres workloads should not be deployed to the three Kubernetes nodes where the postgres workloads live?
That means that it is also recommended to reseize these Kubernetes worker nodes differently from the others, and also have only one postgres cluster for multiple (or all) apps inside a Kubernetes cluster?

**Comment by gbartolini** [comment]
> [@gbartolini](https://github.com/gbartolini) Thank you for your reply. If you need anything else from me just let me know. I believe you saw also the other ticket with a feature request [#6168](https://github.com/cloudnative-pg/cloudnative-pg/issues/6168) , because from my lasts tests I believe that the reason for the failover taking so long is because the failover on case of node failure is dependent on the **node-monitor-grace-period** parameter which is by default in Kubernetes 40s and cannot be changed in a supported way.

Well ... we are talking about a Kubernetes node failure. And our priority is to ensure that there is a consistent view in Kubernetes of not just the cluster, but everything around it, starting from services (through which apps connect). You don't want Kubernetes and the failover manager to have different views on the cluster (this is why we don't have a failover manager in CNPG).

Can you provide more information on the environment you have used for these tests and the actual steps to reproduce the issue exactly?

> Actually putting the controller in the same node as a postgres workloads was only an early experiment, all the later failovertests were executed, when the controller pod was deployed in another node.

Ok. Thanks.

> Bur are you saying that not only the controller pod, but also all workloads other than the postgres workloads should not be deployed to the three Kubernetes nodes where the postgres workloads live? That means that it is also recommended to reseize these Kubernetes worker nodes differently from the others, and also have only one postgres cluster for multiple (or all) apps inside a Kubernetes cluster?

Yes. That is in line with our principle of "cattle vs pets vs elephant herds". PostgreSQL must be treated in a kind of special way, if you want to have best results. Dedicating nodes for Postgres workloads shouldn't be a problem today, thanks to logical (affinity) and physical (taints and labels) separation of workloads that makes them more similar to cattle than pets. This opens up for bare metal Postgres nodes with directly attached storage.

I covered this in my last blog article: https://www.cncf.io/blog/2024/11/20/cloud-neutral-postgres-databases-with-kubernetes-and-cloudnativepg/

Thanks,
Gabriele

**Comment by albionb96** [comment]
@gbartolini I am running openshift 4.15 on prem on top of vSphere cluster (vmware), I have multiple worker nodes and 3 master nodes all of them are in the end effect Virtual Machines on vSphere.

1) I installed the latest cloudnativepg operator from the openshift operator market, and deployed a postgres cluster without changing much from the default configuration., except that I set the tolerationSeconds parameter to 0, since I don't want to wait another additional 300sec before the failover starts, that's documented here: https://cloudnative-pg.io/documentation/1.20/failure_modes/#:~:text=will%20happen%20after-,tolerationSeconds,-.

2) In the end effect I have a postgres cluster with three replicas each of them deployed in different worker nodes (on those worker nodes there are still other workloads which are not related to postgres), and the controller manager pod of cloudnativepg is running on a fourth node.
3) After everything is up and running I start my simple client pod which enters data (e.g. timestamp info) to database each second.
4) After sometime I look in which worker node the primary replica of postgres cluster is living and I go on my vSphere cluster and shut-down exactly that virtualmachine which represents the worker node which hosts the primary replica.
5) It takes something between 40-60seconds until my client is again able to write into it.
6) As explained in this ticket and other linked tickets, I was assuming that this time is coming from the **node-monitor-grace-period** parameter which is by default in Openshift 40s and cannot be changed in a supported way. That's why I made a test and changed this parameter with an unsupportedConfigOverride

$ oc edit kubecontrollermanagers cluster -n openshift-config spec: unsupportedConfigOverrides: extendedArguments: node-monitor-grace-period: - 20s

After this change the failover time was ca. 20s smaller.

Kubernetes may have its own reasons why it waits by default 40s+5mins until the pods are notified that their node is gone. But if we want to have an application which is high available then those default values can never be true. We are using for example NATS as message broker they are also very cloud native from very beginning and they have a very fast failover time, based on their own timeouts. They have also a kind of controller and as soon as the controller notices that a replica is not healthy it will initiate some changes. They are not building the failover mechanism based on what some Kubernetes standards try to introduce as a silver bullet. The custom timeout would be very similar to readinessProbe or livnessProbe which are faster than the node failure mechanism for now.

You said the cloudnativepg is special and should be treated special, why should it be different this time, why should we use only the default solution that Kubernetes offers us in this particular case?

**Comment by albionb96** [comment]
@gbartolini @mnencia @leonardoce Are there any new updates regarding this issue?

**Comment by gbartolini** [comment]
Relying on Kubernetes for consensus is a fundamental design choice that sets CloudNativePG apart from other operators, such as Patroni, which depend on a proxy application to manage the cluster. Given our current commitment, I do not anticipate any changes to this decision in the near future. However, we welcome contributions from others who wish to explore this patch and are happy to review them.

If enabling the configuration of specific timeouts is necessary, we are open to incorporating this into the operator's configuration, similar to what was done in issue #7035.

As we approach the release of version 1.26, we have been working on several other important features and we need to postpone this for now. What do other maintainers think overall?

**Comment by jeremych1000** [comment]
I'm seeing inconsistent failover times of my different CNPG clusters which is very weird. I have two clusters of the same config, and one fails over almost immediately (within 5s), but the other takes a minute or two as seen from the others on this thread.

Not sure how to debug this? Logs are consistent with the above users.

**Comment by i-mighty** [comment]
This is still an issue. I am getting inconsistent failover times of my different CNPG clusters. 

Are there any other known solutions that can be explored?

**Comment by NickSmyr** [comment]
I am. getting these results: 

```
Primary failure (3 nodes): 2 mins downtime
Primary failure (2 nodes): 7 mins downtime
Replica failure (3 nodes): no downtime
Replica failure (2 nodes): 5 mins downtime
``` 

with the following synchronous configuration: 

```
postgresql:
    synchronous:
      method: any
      number: 1
      dataDurability: preferred
```

**Comment by albionb96** [comment]
Hi @gbartolini — I recently came across your blog post [“CNPG Recipe #21 — Finer Control of Postgres Clusters with Liveness Probes”](https://www.gabrielebartolini.it/articles/2025/08/cnpg-recipe-21-finer-control-of-postgres-clusters-with-liveness-probes/?utm_source=chatgpt.com)
 and I was wondering if those enhancements might help with the issue being discussed here.

I know although that these two issues might not be very related, because my understanding is that the liveness probe improvements would cause Kubernetes to restart a failing or isolated primary, which is useful. However, the focus of this ticket is on accelerating the promotion of a replica to primary in cases where the master is down or unreachable.

But still wanted to ask: Do you see overlap here, or would the liveness‑probe logic need to be complemented by additional operator‑level mechanisms to achieve faster failover?
